### ISSUE TITLE
Academy Max Step Set to zero ?

### ISSUE THREAD

**Issue by NonoLG** [Unity-Technologies/ml-agents#2229]
Dear all,

I am currently using ML-agent in order to make my agent to reach targets in different situation.
Until now, I left my Academy Max Step to zero and my agent was apparently learning since I got good results. Nonetheless, I do not really understand the way it works.

Indeed, if my Academy Max Step is set to 0, how are calculated the Episode length ?
What exactly is the mean reward calculated in the Terminal and in Tensorboard ?

Thanks in advance for you answers.

**Comment by gzrjzcx** [comment]
The max step for Academy is the total steps for each episode. If it reaches the max step, then Academy will reset the env. If it set to 0, I guess it means no reset?

**Comment by NonoLG** [comment]
Yes, if It is set to zero, that means no reset.

So, my questions are:

1 - Is the cumulative reward calculated at each step (as a Time Differential Learning) ?

2 - What is the role of the episode ? In my case I randomise parameters each time my agent reach the target or the obstacle.

3 - What is the meaning of the "Episode Length" that we can see in Tensorboard ?

**Comment by roboserg** [comment]
1 - You can print the cum. reward with GetCumulativeReward(). It seems to be calculating at each step. 

2 - the episode is the games episode, its like "game over" and the environment resets

3 - Episode Length is the length in steps of the episode - the time from beginning to the "game over". 

For simple environments you can leave Episode Length at 0 (being infinite) but reset the game objects in the agent itself.

**Comment by NonoLG** [comment]
So, I finally print the Cumulative Reward. It is calculated between the moment the Agent start until it is done. Then it is set to 0.

My remaining questions are:

1 - Is the Tensor board Episode Length in ms ? Indeed, it doesn't correspond to the number of steps (I printed it in same time as the cumulative reward)

2 - Does the Academy Max Step setting impact any Calcul or anything in background (for the ML part for instance), or is it only used to reset the environment if required ?

**Comment by roboserg** [comment]
1 - one unit seems to be 5 steps. I set max. steps to 4000 and tensorboard shows 8000.

2 - cant say for sure. You can train with max step = 0. I only use non 0 if I have many agents and want to reset the whole game. If I have only 1 agent I use max step = 0 and reset the things in the agent.

**Comment by NonoLG** [comment]
Thanks Roboserg. I quite agree with you for the 5 steps. However, I do not understand what is the Steps in the consol

**Comment by roboserg** [comment]
Which steps in the console? There are 60 (I think) unity steps per second. In each unity step functions such as Update() or AgentAction() etc take place. With Time.deltatime you can see how long each step takes.

**Comment by NonoLG** [comment]
I mean:

"INFO:mlagents.trainers: simple3D_8M_MaxAcaSet_V9-0: LearningDiscreteV2: Step: 1000. Mean Reward: 0.400. Std of Reward: 0.917. Training.
INFO:mlagents.trainers: simple3D_8M_MaxAcaSet_V9-0: LearningDiscreteV2: Step: 2000. Mean Reward: 0.250. Std of Reward: 0.873. Training.
INFO:mlagents.trainers: simple3D_8M_MaxAcaSet_V9-0: LearningDiscreteV2: Step: 3000. Mean Reward: 0.450. Std of Reward: 0.907. Training.
INFO:mlagents.trainers: simple3D_8M_MaxAcaSet_V9-0: LearningDiscreteV2: Step: 4000. Mean Reward: 0.100. Std of Reward: 0.800. Training.
INFO:mlagents.trainers: simple3D_8M_MaxAcaSet_V9-0: LearningDiscreteV2: Step: 5000. Mean Reward: -0.100. Std of Reward: 0.663. Training."

What is this STEP ?

**Comment by roboserg** [comment]
Those are the number of training steps in total the agent was trained. It could be millions etc.

**Comment by NonoLG** [comment]
How is that calculated ? I mean, how is that linked to the Agent Step we get by computing GetStepCount ?

Thks

**Comment by gzrjzcx** [comment]
Maybe it just counts the number of steps agent finished. One step for agent is `CollectObs() -> PythonAPI -> Choose Action -> AgentAction() use this action`.

**Comment by roboserg** [comment]
I think its the one step of the environment. However if you run 5 concurrent environments, it will still count 5 environments as one.

**Comment by NonoLG** [comment]
> I think its the one step of the environment. However if you run 5 concurrent environments, it will still count 5 environments as one.

Is "CollectObs() -> PythonAPI -> Choose Action -> AgentAction() use this action." the environment step for you ?

**Comment by NonoLG** [comment]
> 1 - one unit seems to be 5 steps. I set max. steps to 4000 and tensorboard shows ep. len. of 800.
> 
> 2 - cant say for sure. You can train with max step = 0. I only use non 0 if I have many agents and want to reset the whole game. If I have only 1 agent I use max step = 0 and reset the things in the agent.

Regarding the Doc, they explain " A signal that the Episode has ended. The buffer must be reset.
Get only called when the academy resets."

Therefore, what's happening if we leave Academy Max Step to 0 ? How does it learn ?

**Comment by roboserg** [comment]
It learns by collecting buffer_size state action pairs and doing batch_size AFAIK. So that the max step can be 0 and env never reset

**Comment by NonoLG** [comment]
Ok, so the Academy Max Step is only useful if I want to reset different areas at the same time.

In my case, I use to play 8 areas in the same scene. As my target point is reset (random pos) each time my agent reach it, I can keep Academy Max step to zero, even if I have 8 areas ?

Thanks for you answers

**Comment by ervteng** [comment]
Thanks for the comments - closing this issue due to inactivity. Feel free to open a new issue if you're still running into problems.
